be/src/runtime/hdfs/hdfs_fs_cache.h

```diff
@@ -11,22 +13,31 @@
 
 #include "common/status.h"
 #include "env/env_hdfs.h"
+
 namespace starrocks {
 
+class S3Client;
+
+struct HdfsFsHandle {
+    enum class Type { LOCAL, HDFS, S3 };
+    Type type;
+    std::string namenode;
+    hdfsFS hdfs_fs;
+    S3Client* s3_client;
+};
+
 // Cache for HDFS file system
 class HdfsFsCache {
 public:
-    using FileOpenLimit = std::atomic<int32_t>;
-    using FileOpenLimitPtr = FileOpenLimit*;
-    using HdfsFsMap = std::unordered_map<std::string, std::pair<HdfsFsHandle, std::unique_ptr<FileOpenLimit>>>;
+    using HdfsFsMap = std::unordered_map<std::string, HdfsFsHandle>;
 
     static HdfsFsCache* instance() {
         static HdfsFsCache s_instance;
         return &s_instance;
     }
 
     // This function is thread-safe
-    Status get_connection(const std::string& namenode, HdfsFsHandle* handle, FileOpenLimitPtr* semaphore);
+    Status get_connection(const std::string& namenode, HdfsFsHandle* handle);
```

Comment:
Why to remove it ?

Reply:
> Why to remove it ?

I'v moved the logic to a new class `OpenLimitAllocator` in hdfs_scan_node.cpp, because this method will be called from the inside of `EnvHdfs::new_random_access_file` and there is no way to pass the `open_limit` to the outside of `EnvHdfs` without abstraction leak.